### ISSUE TITLE
Random ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR(s) after upgrading to Rails 5.1.1

### ISSUE THREAD

**Issue by kevinsapp** [rails/rails#29164]
### Steps to reproduce

I simply updated from Rails 5.0.3 to 5.1.1 in my project. 
Now, when I run tests while in development mode, I see errors like this:

ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "end_users" does not exist...

### Expected behavior
Expected tests to run without error.

### Actual behavior
Various tests throw errors. I always get an error similar to `ActiveRecord::StatementInvalid: PG::UndefinedTable: ERROR:  relation "end_users" does not exist...`

But, I observed that the errors are not consistent - i.e. during some test runs I see the error on the end_users table, other times I see the error with other tables.

While googling this issue, I found a StackOverflow Q&A which gives me a hint that Spring preloader may be the culprit: [ActiveRecord error when running rails console](http://stackoverflow.com/questions/43863458/activerecord-error-when-running-rails-console)

Simply stopping Spring before running tests gives me inconsistent results (sometimes all tests pass, more often some fail).

However, when I do the following:

1. Open a new terminal
1. Navigate to my project directory
1. Do `spring stop` (returns Spring is not running)
1. Do `export DISABLE_SPRING=1`
1. Do `rails t`

All tests run fine and are green the _first time_ I run them!

If I do `rails t` _again_, within the same terminal some tests now fail (again, different tests fail each time).

Is there some special upgrade step I might have missed that would cause such an error? Is this a known issue with the Spring preloader? Is this a bug in ActiveRecord?

Thank you!

### System configuration
**Rails version**: 5.1.1

**Ruby version**: ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]

**Comment by prathamesh-sonpatki** [comment]
Can you upload a sample application on Github that reproduces the issue? Thanks.

**Comment by mkilling** [comment]
I'm currently investigating a similar problem (same exception) that appeared after upgrading from activerecord 4.2.7 to activerecord 4.2.8.

It seems that in some situations a table that is required is not being joined in in the query. In contrast to @kevinsapp the problem does not disappear after disabling spring. I'm trying to extract a minimal test case but it‘s tricky.

**Comment by mkilling** [comment]
A first step towards isolating this.

I run this code from within a scope (class method):
```
my_query = playtest.game.playtests.joins(participations: :tester)
puts my_query.to_sql
```

With `activerecord-4.2.7` it results in this SQL statement:
```
SELECT "playtests".* FROM "playtests"
INNER JOIN "buckets" ON "buckets"."playtest_id" = "playtests"."id"
INNER JOIN "participations" ON "participations"."bucket_id" = "buckets"."id"
INNER JOIN "users" ON "users"."id" = "participations"."tester_id" AND "users"."type" IN ('Tester')
WHERE "playtests"."game_id" = 2755
```

With `activerecord-4.2.8` the query looks like this (note the additional conditions after `INNER JOIN "users"):
```
SELECT "playtests".* FROM "playtests"
INNER JOIN "buckets" ON "buckets"."playtest_id" = "playtests"."id"
INNER JOIN "participations" ON "participations"."bucket_id" = "buckets"."id"
INNER JOIN "users" ON "users"."id" = "participations"."tester_id" AND "users"."type" IN ('Tester') AND "users"."type" IN ('Tester') AND "interests"."id" = 20 AND (1=1) AND "users"."is_qualified" = 't' AND (1=1) AND "users"."banned" = 'f' AND "users"."is_test_account" = 'f' AND "users"."receives_invitations" = 't' AND "users"."custom_panel_id" IS NULL AND "users"."referring_team_id" IS NULL AND "users"."email_bounced_at" IS NULL
WHERE "playtests"."game_id" = 2751
```

To me it seems like some additional conditions somehow `leak` into this query...

**Comment by mkilling** [comment]
For me, the issue was introduced by rails commit c55a5d6b16d18b028c3642caaf5815072630de40. I‘m investigating further to see if the culprit might be in our code.

**Comment by mkilling** [comment]
I wrote a minimal sample application that illustrates my problem, but I feel it might be a different issue from the one that @kevinsapp is experiencing because my problem is 100% reproducible. https://github.com/rails/rails/issues/29338

**Comment by kevinsapp** [comment]
**Update**

So, I now think this business about the Spring preloader is a red herring.

My guess is that the problem I'm seeing is more probably related to a new feature of Rails 5.1: [Transactional tests with multiple connections](http://edgeguides.rubyonrails.org/5_1_release_notes.html#transactional-tests-with-multiple-connections)

The hint that guided me to that guess is this issue report: https://github.com/rails/rails/issues/28466. If I set `self.use_transactional_tests = false` in my test_helper the particular issue that I reported here goes away.

From what I can tell, this feature was implemented by @eileencodes in this PR: [Ensure test threads share a DB connection](https://github.com/rails/rails/pull/28083). She may be the best person to understand this problem if she has time.

**Context**

My application uses PostgreSQL as the database, and I often need to set the `connection.schema_search_path` so that I can access tables in a particular Postgres schema. The code that does this is usually, but not exclusively, called from an "around filter" in my ApplicationController. If the `connection.schema_search_path` isn't set to the correct schema, then I would expect to see errors finding tables, which is what I'm seeing. 

As I reported earlier, tests seem to fail randomly - i.e. different tests fail each time. Could it be that I've stumbled into a race condition caused by multiple tests attempting to set `connection.schema_search_path` simultaneously on a shared DB connection?

@mkilling I think you're right in that we're probably experiencing different problems.

**Comment by eileencodes** [comment]
@mkilling your issue doesn't sound related so please open a new one.

@kevinsapp do you have an app you can share with us that reproduces the issue? It's hard for us to determine what's wrong without a way to reproduce it.

Are you running multiple tests in parallel / concurrently when you're seeing this happen? How are you setting `connection.schema_search_path` during test runs? The documentation says that should only get set by the database.yml so maybe something odd is happening with the way you're setting the `schema_search_path`.

@matthewd debugged a couple issues with the thread changes we made relating to deadlocks in postgres https://github.com/rails/rails/pull/28083 but I believe this is a new issue.

**Comment by kevinsapp** [comment]
Hi @eileencodes 

Unfortunately, I can't share the actual app that's having these issues. I did build a small sample app which uses PostgreSQL as the database and sets the connection.schema_search_path in several places, much like my real app does. The sample app probably won't be helpful in quickly reproducing the issue because it only has a handful of tests. My real app has a moderate amount of tests (~500), and I find that I'm much more likely to run into problems when running the full test suite vs. a small subset.

If you'd like to take a look, the sample app is here: https://github.com/kevinsapp/buggy

The relevant code is in
`app/models/tenant.rb`
`app/controllers/application_controller.eb`
`test/controllers/users_controller_test.eb`

I'm not intentionally running anything in parallel. When I run tests, I just do `rails t` and that's it. My previous question: "Could it be that I've stumbled into a race condition caused by multiple tests attempting to set connection.schema_search_path simultaneously on a shared DB connection?" was just asking how the Rails test harness works. I occasionally see errors like so: "No live threads left. Deadlock? 5 threads, 5 sleeps current when running tests" when running tests.
